File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
class Solution {
public:
    int n1, n2;

    int fun(int i, int j, string &text1, string &text2, vector<vector<int>> &dp) {
        if(i >= n1 || j >= n2) return 0;

        if(dp[i][j] != -1) return dp[i][j];

        if(text1[i] == text2[j]) {
            return dp[i][j] = 1 + fun(i + 1, j + 1, text1, text2, dp);
        }

        return dp[i][j] = max(
            fun(i + 1, j, text1, text2, dp),
            fun(i, j + 1, text1, text2, dp)
        );
    }

    int longestCommonSubsequence(string text1, string text2) {
        n1 = text1.size();
        n2 = text2.size();

        vector<vector<int>> dp(n1, vector<int>(n2, -1));

        return fun(0, 0, text1, text2, dp);
    }
};

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.py
class Solution:
    def fun(self, i, j, text1, text2,dp):

        if i == len(text1) or j == len(text2):
            return 0
        if dp[i][j]!=-1 :
            return dp[i][j]
        
        if text1[i] == text2[j]:
            dp[i][j]= 1 + self.fun(i + 1, j + 1, text1, text2,dp)
            return dp[i][j]

        else:
            dp[i][j]=max(
                self.fun(i + 1, j, text1, text2,dp),
                self.fun(i, j + 1, text1, text2,dp)
            )
            return dp[i][j]

    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        r=len(text1)
        c=len(text2)

        dp=[[-1]*(c+1) for _ in range(r+1)]

        return self.fun(0, 0, text1, text2,dp)

